File: app/src/test/java/com/peterlaurence/trekme/core/map/MapParserTest.java
package com.peterlaurence.trekme.core.map;

import com.peterlaurence.trekme.BuildConfig;
import com.peterlaurence.trekme.core.map.gson.MarkerGson;
import com.peterlaurence.trekme.core.map.gson.RouteGson;
import com.peterlaurence.trekme.core.map.maploader.MapLoader;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.io.File;
import java.net.URL;
import java.util.List;

import static org.junit.Assert.assertEquals;

/**
 * Unit tests for maps's json file parsing.
 *
 * @author peterLaurence on 26/02/17.
 */
@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class)
public class MapParserTest {
    private static File mJsonFilesDirectory;

    static {
        try {
            URL mapDirURL = MapImporterTest.class.getClassLoader().getResource("mapjson-example");
            mJsonFilesDirectory = new File(mapDirURL.toURI());
        } catch (Exception e) {
            System.out.println("No json directory found.");
        }
    }

    @Test
    public void mapTracksParse() {
        if (mJsonFilesDirectory != null) {
            File[] dirs = new File[1];
            dirs[0] = mJsonFilesDirectory;
            final Map[] map = new Map[1];

            MapLoader.MapListUpdateListener mapListUpdateListener = new MapLoader.MapListUpdateListener() {
                @Override
                public void onMapListUpdate(boolean mapsFound) {
                    List<Map> mapList = MapLoader.getInstance().getMaps();

                    /* One map should be found */
                    assertEquals(1, mapList.size());
                    map[0] = mapList.get(0);
                }
            };

            MapLoader.MapRouteUpdateListener mapRouteUpdateListener = new MapLoader.MapRouteUpdateListener() {
                @Override
                public void onMapRouteUpdate() {
                    /* 2 routes should be found */
                    assertEquals(2, map[0].getRoutes().size());

                    RouteGson.Route route = map[0].getRoutes().get(0);
                    assertEquals("A test route 1", route.name);
                    assertEquals(true, route.visible);
                    List<MarkerGson.Marker> markers = route.route_markers;
                    assertEquals(2, markers.size());

                    MarkerGson.Marker marker1 = markers.get(0);
                    assertEquals("marker1", marker1.name);
                    assertEquals(6198798.5047565, marker1.proj_x, 0);

                    MarkerGson.Marker marker2 = markers.get(1);
                    assertEquals("marker2", marker2.name);
                    assertEquals(-2418744.7142449305, marker2.proj_y, 0);
                }
            };

            MapLoader mapLoader = MapLoader.getInstance();
            mapLoader.setMapRouteUpdateListener(mapRouteUpdateListener);
            mapLoader.setMapListUpdateListener(mapListUpdateListener);
            mapLoader.clearAndGenerateMaps(dirs);
            mapLoader.getRoutesForMap(map[0]);
        }
    }
}

File: app/src/test/java/com/peterlaurence/trekme/core/map/MapParserTest.kt
package com.peterlaurence.trekme.core.map

import com.peterlaurence.trekme.core.map.data.dao.FileBasedMapRegistry
import com.peterlaurence.trekme.core.map.data.dao.MapLoaderDaoFileBased
import com.peterlaurence.trekme.core.map.data.dao.MapSaverDaoImpl
import com.peterlaurence.trekme.core.map.data.dao.RouteDaoImpl
import com.peterlaurence.trekme.core.map.di.MapModule
import com.peterlaurence.trekme.core.map.domain.repository.RouteRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.io.File
import kotlin.test.assertNotNull

/**
 * Unit tests for maps's json file parsing.
 *
 * @since 2017/02/26 -- Converted to Kotlin on 2019/02/17
 */
@RunWith(RobolectricTestRunner::class)
class MapParserTest {
    private val registry = FileBasedMapRegistry()
    private val gson = MapModule.provideGson()
    private val mapSaverDao = MapSaverDaoImpl(registry, Dispatchers.Unconfined, Dispatchers.IO, gson)
    private val mapLoaderDao = MapLoaderDaoFileBased(
        registry, mapSaverDao, gson, MapModule.provideJson(), Dispatchers.IO
    )

    private val routeDao = RouteDaoImpl(registry, Dispatchers.IO, Dispatchers.Unconfined)
    private val routeRepository = RouteRepository(routeDao)

    @Test
    fun legacyMapRoutesParse() = runBlocking {
        val mapDirURL =
            MapImportInteractorTest::class.java.classLoader!!.getResource("map-with-legacy-routes")
        val mapDir = File(mapDirURL.toURI())

        assertTrue(mapDir.exists())

        val dirs = listOf(mapDir)
        val mapList = mapLoaderDao.loadMaps(dirs)

        /* One map should be found */
        assertEquals(1, mapList.size.toLong())
        val map = mapList[0]

        routeRepository.importRoutes(map)
        assertEquals(2, map.routes.value.size.toLong())

        val route = map.routes.value[0]
        assertEquals("A test route 1", route.name)
        assertTrue(route.visible.value)
        val markers = route.routeMarkers
        assertEquals(2, markers.size.toLong())

        val marker1 = markers[0]
        assertEquals("marker1", marker1.name)
        val lat = marker1.lat
        assertNotNull(lat)
        assertEquals(0.0, lat, 0.0)

        val marker2 = markers[1]
        assertEquals("marker2", marker2.name)
        val lon = marker2.lon
        assertNotNull(lon)
        assertEquals(0.0, lon, 0.0)
    }

    @Test
    fun mapRoutesParse() = runBlocking {
        val mapDirURL = MapImportInteractorTest::class.java.classLoader!!.getResource("map-with-routes")
        val mapDir = File(mapDirURL.toURI())

        assertTrue(mapDir.exists())

        val dirs = listOf(mapDir)
        val mapList = mapLoaderDao.loadMaps(dirs)

        /* One map should be found */
        assertEquals(1, mapList.size.toLong())
        val map = mapList[0]

        routeRepository.importRoutes(map)
        assertEquals(2, map.routes.value.size.toLong())

        val route1 = map.routes.value.sortedBy { it.name }[0]
        assertEquals("track-10-09-2021_22h09-04s", route1.name)
        assertTrue(route1.visible.value)
        val markers1 = route1.routeMarkers
        assertEquals(2, markers1.size.toLong())

        val marker1 = markers1[0]
        assertEquals("marker1", marker1.name)
        val lat = marker1.lat
        assertNotNull(lat)
        assertEquals(40.1, lat, 0.0)

        val marker2 = markers1[1]
        assertEquals("marker2", marker2.name)
        val lon = marker2.lon
        assertNotNull(lon)
        assertEquals(12.57, lon, 0.0)

        val route2 = map.routes.value.sortedBy { it.name }[1]
        assertEquals("track-13-07-2021_10h27-44s", route2.name)
        assertTrue(route2.visible.value)
        val markers2 = route2.routeMarkers
        assertEquals(2, markers2.size.toLong())

        val marker3 = markers2[0]
        assertEquals("marker3", marker3.name)
        val lat2 = marker3.lat
        assertNotNull(lat2)
        assertEquals(12.3, lat2, 0.0)

        val marker4 = markers2[1]
        assertEquals("marker4", marker4.name)
        val lon2 = marker4.lon
        assertNotNull(lon2)
        assertEquals(22.8, lon2, 0.0)
    }
}
